samples/tooling/profiles/profiles.cpp

```diff
@@ -0,0 +1,711 @@
+	camera.set_rotation(glm::vec3(0.0f));
+
+	// Note: Using reversed depth-buffer for increased precision, so Znear and Zfar are flipped
+	camera.set_perspective(60.0f, (float) width / (float) height, 256.0f, 0.1f);
+
+	generate_textures();
+	generate_cubes();
+	prepare_uniform_buffers();
+	setup_descriptor_set_layout();
+	prepare_pipelines();
+	setup_descriptor_pool();
+	setup_descriptor_set();
+	build_command_buffers();
+	prepared = true;
+	return true;
+}
+
+void Profiles::render(float delta_time)
+{
+	if (!prepared)
+		return;
+	draw();
+}
+
+void Profiles::view_changed()
+{
+	update_uniform_buffers();
+}
+
+// This sample overrides the device creation part of the framework
+// Instead of manually setting up all extensions, features, etc. we use the Vulkan Profiles library to simplify device setup
+void Profiles::create_device()
+{
+	auto &gpu = instance->get_suitable_gpu(surface);
+
+	// Simplified queue setup (only graphics)
+	uint32_t                selected_queue_family   = 0;
+	const auto &            queue_family_properties = gpu.get_queue_family_properties();
+	const float             default_queue_priority{0.0f};
+	VkDeviceQueueCreateInfo queue_create_info{};
+	queue_create_info.sType            = VK_STRUCTURE_TYPE_DEVICE_QUEUE_CREATE_INFO;
+	queue_create_info.queueCount       = 1;
+	queue_create_info.pQueuePriorities = &default_queue_priority;
+	for (uint32_t i = 0; i < static_cast<uint32_t>(queue_family_properties.size()); i++)
+	{
+		if (queue_family_properties[i].queueFlags & VK_QUEUE_GRAPHICS_BIT)
+		{
+			queue_create_info.queueFamilyIndex = i;
+			selected_queue_family              = i;
+			break;
+		}
+	}
+
+	VkDeviceCreateInfo create_info{VK_STRUCTURE_TYPE_DEVICE_CREATE_INFO};
+	create_info.pNext                = gpu.get_extension_feature_chain();
+	create_info.pQueueCreateInfos    = &queue_create_info;
+	create_info.queueCreateInfoCount = 1;
+
+	const VpProfileProperties profile_properties = {PROFILE_NAME, PROFILE_SPEC_VERSION};
+
+	// Check if the profile is supported at device level
+	VkBool32 profile_supported;
+	vpGetPhysicalDeviceProfileSupport(instance->get_handle(), gpu.get_handle(), &profile_properties, &profile_supported);
+	if (!profile_supported)
+	{
+		throw std::runtime_error{"The selected profile is not supported (error at creating the device)!"};
+	}
+
+	// Create the device using the profile tool library
+	VpDeviceCreateInfo deviceCreateInfo{};
+	deviceCreateInfo.pCreateInfo = &create_info;
+	deviceCreateInfo.pProfile    = &profile_properties;
+	deviceCreateInfo.flags       = VP_DEVICE_CREATE_MERGE_EXTENSIONS_BIT;
+	VkDevice vulkan_device;
+	VkResult result = vpCreateDevice(gpu.get_handle(), &deviceCreateInfo, nullptr, &vulkan_device);
+
+	if (result != VK_SUCCESS)
+	{
+		throw vkb::VulkanException{result, "Could not create device with the selected profile. The device may not support all features required by this profile!"};
+	}
+
+	// Post device setup required for the framework
```

Comment:
for my benefit, why are these new helpers needed now? do they duplicate framework code?

Reply:
No duplication. The code in those helpers was part of the base class device setup and not accessible from the outside. As I didn't want to duplicate that whole code in the sample, I moved the code into those helpers and use them both in the framework device class as well as the sample.